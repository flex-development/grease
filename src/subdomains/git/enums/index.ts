/**
 * @file Entry Point - Enums
 * @module grease/git/enums
 */

export { default as RepositoryProvider } from './repository-provider'
