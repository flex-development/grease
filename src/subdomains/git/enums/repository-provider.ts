/**
 * @file Enums - RepositoryProvider
 * @module grease/git/enums/RepositoryProvider
 */

/**
 * Repository provider names.
 *
 * @enum {Lowercase<string>}
 */
enum RepositoryProvider {
  BITBUCKET = 'bitbucket',
  GITHUB = 'github',
  GITLAB = 'gitlab'
}

export default RepositoryProvider
