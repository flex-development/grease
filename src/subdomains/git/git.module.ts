/**
 * @file GitModule
 * @module grease/git/GitModule
 */

import { LoggerService, ValidationService } from '#src/providers'
import { Global, Module } from '@nestjs/common'
import { GitService } from './providers'

/**
 * Global git operations module.
 *
 * @class
 */
@Global()
@Module({
  exports: [GitService],
  providers: [GitService, LoggerService, ValidationService]
})
class GitModule {}

export default GitModule
