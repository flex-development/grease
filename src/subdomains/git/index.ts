/**
 * @file Entry Point - git
 * @module grease/git
 */

export * from './enums'
export { default as GitModule } from './git.module'
export type * from './interfaces'
export * from './models'
export * from './options'
export * from './providers'
export type * from './types'
