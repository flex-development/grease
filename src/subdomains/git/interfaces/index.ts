/**
 * @file Entry Point - Interfaces
 * @module grease/git/interfaces
 */

export type { default as ICommit } from './commit.interface'
