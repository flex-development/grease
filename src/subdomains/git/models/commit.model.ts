/**
 * @file Models - Commit
 * @module grease/git/models/Commit
 */

import type { RepositoryProvider } from '#src/git/enums'
import type { ICommit } from '#src/git/interfaces'
import type { CommitGrammarOptions } from '#src/git/options'
import type {
  Author,
  BreakingChange,
  CommitFields,
  Reference,
  Trailer
} from '#src/git/types'
import {
  at,
  cast,
  construct,
  defaults,
  get,
  ifelse,
  objectify,
  pick,
  select,
  split,
  trim,
  trimEnd,
  unique,
  type Nullable,
  type Simplify
} from '@flex-development/tutils'
import CommitGrammar from './commit-grammar.model'
import Repository from './repository.model'

/**
 * Parsed commit model.
 *
 * @class
 * @implements {ICommit}
 */
class Commit implements ICommit {
  /**
   * Raw commit chunk.
   *
   * @protected
   * @readonly
   * @instance
   * @member {string} chunk
   */
  protected readonly chunk: string

  /**
   * Raw commit fields object.
   *
   * @see {@linkcode CommitFields}
   *
   * @protected
   * @readonly
   * @instance
   * @member {CommitFields} fields
   */
  protected readonly fields: CommitFields

  /**
   * Commit grammar instance.
   *
   * @see {@linkcode CommitGrammar}
   *
   * @public
   * @readonly
   * @instance
   * @member {CommitGrammar} grammar
   */
  public readonly grammar: CommitGrammar

  /**
   * Repository provider name.
   *
   * @public
   * @readonly
   * @instance
   * @member {RepositoryProvider} provider
   */
  public readonly provider: RepositoryProvider

  /**
   * Parent repository.
   *
   * @public
   * @readonly
   * @instance
   * @member {Repository} repository
   */
  public readonly repository: Repository

  /**
   * Create a new parsed commit.
   *
   * @param {Buffer | Commit | string} chunk - Raw commit chunk or parsed commit
   * @param {Repository?} [repository=new Repository()] - Parent repository
   * @param {Partial<CommitGrammarOptions>?} [grammar] - Commit grammar options
   */
  constructor(
    chunk: Buffer | Commit | string,
    repository: Repository = new Repository(),
    grammar?: Partial<CommitGrammarOptions>
  ) {
    this.chunk = (chunk = chunk.toString()).padEnd(chunk.length + 1, '\n')
    if (!trim(this.chunk)) throw new Error('empty commit chunk')

    // set repository and provider
    this.repository = get(chunk, 'repository', repository)
    this.provider = this.repository.provider

    // set commit grammar
    this.grammar = get(chunk, 'grammar', new CommitGrammar({
      ...grammar,
      provider: this.provider
    }))

    // init commit fields object
    this.fields = construct(cast<CommitFields>(objectify(
      [...this.chunk.matchAll(this.grammar.field)],
      match => get(match, 'groups.field', ''),
      match => trim(get(match, 'groups.value', ''))
    )))

    // add header data to commit fields object
    this.fields = defaults(
      this.fields,
      pick(get(this.grammar.header.exec(this.fields.header), 'groups')!, [
        'breaking',
        'pr',
        'scope',
        'subject',
        'type'
      ]),
      {
        breaking: '',
        pr: 'null',
        scope: '',
        subject: '',
        type: ''
      }
    )

    // remove extra chunk padding
    this.chunk = this.chunk.slice(0, -1)
  }

  /**
   * Commit author details.
   *
   * @see {@linkcode Author}
   *
   * @public
   *
   * @return {Author} Commit author object
   */
  public get author(): Author {
    return this.fields.author
  }

  /**
   * Commit body text.
   *
   * @public
   *
   * @return {string} Commit body
   */
  public get body(): string {
    return trimEnd(this.fields.body.replace(this.fields.trailers, ''))
  }

  /**
   * Commit contains breaking changes?
   *
   * @public
   *
   * @return {boolean} Boolean indicating commits contains breaking changes
   */
  public get breaking(): boolean {
    return this.breaks.length > 0
  }

  /**
   * Breaking changes noted in {@linkcode subject} and {@linkcode trailers}.
   *
   * @see {@linkcode BreakingChange}
   *
   * @public
   *
   * @return {BreakingChange<Commit>[]} Breaking changes array
   */
  public get breaks(): BreakingChange<Commit>[] {
    return unique(select(
      [
        this.fields.breaking
          ? {
            pr: this.pr,
            scope: this.scope,
            token: 'BREAKING-CHANGE',
            type: this.type,
            value: this.subject
          }
          : <Trailer>{ token: this.fields.breaking },
        ...this.trailers
      ],
      trailer => /^BREAKING[ -]CHANGE/.test(trailer.token),
      trailer => ({
        commit: this,
        mentions: this.mentions,
        pr: get(trailer, 'pr', null),
        scope: get(trailer, 'scope', null),
        subject: trailer.value,
        type: this.type
      })
    ))
  }

  /**
   * Commit date in strict ISO 8601 format (`%cI`).
   *
   * @see https://git-scm.com/docs/pretty-formats/2.21.0
   *
   * @public
   *
   * @return {string} Commit date
   */
  public get date(): string {
    return this.fields.date
  }

  /**
   * Abbreviated commit {@linkcode sha}.
   *
   * @public
   *
   * @return {string} Commit hash
   */
  public get hash(): string {
    return this.fields.hash
  }

  /**
   * Commit {@linkcode type}, {@linkcode scope}, breaking change indicator, and
   * {@linkcode subject}.
   *
   * @public
   *
   * @return {string} Commit header text
   */
  public get header(): string {
    return this.fields.header
  }

  /**
   * Users mentioned in commit.
   *
   * @public
   *
   * @return {string[]} Mentions array
   */
  public get mentions(): string[] {
    return select(
      [...this.chunk.matchAll(this.grammar.mention)],
      null,
      match => get(match, 'groups.user')!
    )
  }

  /**
   * Pull request number.
   *
   * @public
   *
   * @return {Nullable<number>} Pull request number or `null`
   */
  public get pr(): Nullable<number> {
    return cast(JSON.parse(this.fields.pr))
  }

  /**
   * Issue and pull request references.
   *
   * @see {@linkcode Reference}
   *
   * @public
   *
   * @return {Reference[]} Reference array
   */
  public get references(): Reference[] {
    return select(
      [...this.chunk.matchAll(this.grammar.reference)],
      null,
      m => ({
        action: get(m, 'groups.action', null),
        number: JSON.parse(get(m, 'groups.number', 'null')),
        owner: get(m, 'groups.owner', null),
        ref: get(m, 'groups.ref', ''),
        repo: get(m, 'groups.repo', null)
      })
    )
  }

  /**
   * Commit scope.
   *
   * @public
   *
   * @return {Nullable<string>} Commit scope or `null`
   */
  public get scope(): Nullable<string> {
    return ifelse(this.fields.scope, this.fields.scope, null)
  }

  /**
   * Commit sha.
   *
   * @public
   *
   * @return {string} Commit sha
   */
  public get sha(): string {
    return this.fields.sha
  }

  /**
   * Commit subject.
   *
   * @public
   *
   * @return {string} Commit subject text
   */
  public get subject(): string {
    return this.fields.subject
  }

  /**
   * Tags associated with commit.
   *
   * @public
   *
   * @return {string[]} Tags array
   */
  public get tags(): string[] {
    return select(
      select(
        split(this.fields.tags, ','),
        tag => !!tag.length,
        tag => trim(tag.replace(/^tag: */, ''))
      ),
      tag => this.grammar.tag.test(tag)
    )
  }

  /**
   * Commit message trailers.
   *
   * @public
   *
   * @return {Trailer[]} Trailers array
   */
  public get trailers(): Trailer[] {
    return select(
      [...this.fields.trailers.matchAll(this.grammar.trailer)],
      null,
      match => pick(match.groups!, ['token', 'value'])
    )
  }

  /**
   * Commit type.
   *
   * @public
   *
   * @return {string} Commit type
   */
  public get type(): string {
    return this.fields.type
  }

  /**
   * Tagged commit version.
   *
   * @public
   *
   * @return {Nullable<string>} First tag in {@linkcode tags} or `null`
   */
  public get version(): Nullable<string> {
    return at(this.tags, 0, null)
  }

  /**
   * Get a JSON-serializable commit object.
   *
   * @public
   *
   * @return {Simplify<ICommit>} JSON-serializable commit
   */
  public toJSON(): Simplify<ICommit> {
    return {
      author: this.author,
      body: this.body,
      breaking: this.breaking,
      breaks: this.breaks,
      date: this.date,
      hash: this.hash,
      header: this.header,
      mentions: this.mentions,
      pr: this.pr,
      references: this.references,
      scope: this.scope,
      sha: this.sha,
      subject: this.subject,
      tags: this.tags,
      trailers: this.trailers,
      type: this.type,
      version: this.version
    }
  }

  /**
   * Get `this` raw commit.
   *
   * @public
   *
   * @return {string} Raw commit string
   */
  public toString(): string {
    return this.chunk
  }
}

export default Commit
