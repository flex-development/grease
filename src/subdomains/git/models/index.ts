/**
 * @file Entry Point - Models
 * @module grease/git/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Repository } from './repository.model'
