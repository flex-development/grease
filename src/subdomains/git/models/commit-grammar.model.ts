/**
 * @file Models - CommitGrammar
 * @module grease/git/models/CommitGrammar
 */

import { CommitGrammarOptions } from '#src/git/options'
import {
  join,
  regexp,
  select,
  template,
  unique,
  type Partial
} from '@flex-development/tutils'

/**
 * Commit parser grammar.
 *
 * @class
 */
class CommitGrammar {
  /**
   * Grammar options.
   *
   * @see {@linkcode CommitGrammarOptions}
   *
   * @protected
   * @readonly
   * @instance
   * @member {CommitGrammarOptions} options
   */
  protected readonly options: CommitGrammarOptions

  /**
   * Create a new commit grammar instance.
   *
   * @see {@linkcode CommitGrammarOptions}
   *
   * @param {Partial<CommitGrammarOptions>?} [options] - Grammar options
   */
  constructor(options?: Partial<CommitGrammarOptions>) {
    this.options = new CommitGrammarOptions(options)
  }

  /**
   * Regular expression matching fields in a raw commit.
   *
   * @see https://regex101.com/r/QnFMsV
   *
   * @public
   *
   * @return {RegExp} Raw commit field regex
   */
  public get field(): RegExp {
    return /^-(?<field>[^\s$]+?)-\n(?<value>.*?(?=\n*^-\S+?-)|(?:.*$))/gms
  }

  /**
   * Regular expression matching a commit header.
   *
   * @see https://regex101.com/r/WAJeLp
   * @see https://conventionalcommits.org
   *
   * @public
   *
   * @return {RegExp} Commit header regex
   */
  public get header(): RegExp {
    return /^(?<type>[a-z]+)(?:\((?<scope>[a-z-]+)\))?(?<breaking>!)?: +(?<subject>(?:.+ \([!#](?<pr>\d+)\))|.+)/i
  }

  /**
   * Regular expression matching mentions (e.g. `@unicornware`) in a raw commit.
   *
   * @see https://regex101.com/r/upbRpj
   *
   * @public
   *
   * @return {RegExp} Mention regex
   */
  public get mention(): RegExp {
    return /\B(?<mention>@(?<user>[\w-]{1,38}(?=[^\w-]\B|\s)))/g
  }

  /**
   * Get a regular expression matching references in a raw commit.
   *
   * @see https://regex101.com/r/Thsp1M
   *
   * @public
   *
   * @return {RegExp} Reference regex
   */
  public get reference(): RegExp {
    /**
     * Pattern matching issue references in commit bodies.
     *
     * @const {string} pattern
     */
    const pattern: string =
      '(?:(?:(?<action>{actions}) +)|(?<repository>(?<owner>[\\da-z](?:-(?=[\\da-z])|[\\da-z]){0,38}(?<=[\\da-z]))\\/(?<repo>\\S+)))?(?<ref>(?<prefix>{prefixes})(?<number>\\d+))'

    return new RegExp(
      template(pattern, {
        actions: join(this.options.actions, '|'),
        prefixes: join(select(unique([
          ...this.options.issues,
          ...this.options.pr
        ]), null, regexp), '|')
      }),
      'gi'
    )
  }

  /**
   * Regular expression matching a semantic release tag.
   *
   * @see https://regex101.com/r/WyQyHZ
   * @see https://semver.org
   *
   * @public
   *
   * @return {RegExp} Semantic release tag regex
   */
  public get tag(): RegExp {
    /**
     * Pattern matching semantic release tags.
     *
     * @const {string} pattern
     */
    const pattern: string =
      '^(?<prefix>{prefix})(?<version>(?<major>0|[1-9]\\d*)\\.(?<minor>0|[1-9]\\d*)\\.(?<patch>0|[1-9]\\d*)(?:-(?<prerelease>(?:0|[1-9]\\d*|\\d*[A-Za-z-][\\dA-Za-z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[A-Za-z-][\\dA-Za-z-]*))*))?(?:\\+(?<metadata>[\\dA-Za-z-]+(?:\\.[\\dA-Za-z-]+)*))?)$'

    return new RegExp(template(pattern, { prefix: this.options.tagprefix }))
  }

  /**
   * Regular expression matching trailers in a raw commit.
   *
   * @see https://git-scm.com/docs/git-interpret-trailers
   * @see https://regex101.com/r/I56Kgg
   *
   * @return {RegExp} Git trailer regex
   */
  public get trailer(): RegExp {
    return /(?<=^|\n)(?<token>[\w -]+?(?=:)): (?<value>[\S\s]+?(?:(?=\n[\w -]+?(?=:))|(?=\n*$)))/g
  }
}

export default CommitGrammar
