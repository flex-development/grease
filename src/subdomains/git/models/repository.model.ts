/**
 * @file Models - Repository
 * @module grease/git/models/Repository
 */

import { RepositoryProvider } from '#src/git/enums'
import type {
  RepositoryHost,
  RepositoryKeywords
} from '#src/git/types'
import { PackageManifest } from '#src/models'
import {
  cast,
  defaults,
  get,
  ifelse,
  template,
  trim
} from '@flex-development/tutils'
import {
  IsEnum,
  IsString,
  MinLength
} from 'class-validator'

/**
 * Repository model.
 *
 * @class
 */
class Repository {
  /**
   * Repository owner.
   *
   * @example
   *  'flex-development'
   *
   * @public
   * @readonly
   * @instance
   * @member {string} owner
   */
  @IsString()
  @MinLength(1)
  public readonly owner: string

  /**
   * Repository provider name.
   *
   * @see {@linkcode RepositoryProvider}
   *
   * @default RepositoryProvider.GITHUB
   *
   * @public
   * @readonly
   * @instance
   * @member {RepositoryProvider} provider
   */
  @IsEnum(RepositoryProvider)
  public readonly provider: RepositoryProvider

  /**
   * Repository name.
   *
   * @example
   *  'grease'
   *
   * @public
   * @readonly
   * @instance
   * @member {string} repo
   */
  @IsString()
  @MinLength(1)
  public readonly repo: string

  /**
   * Create a new repository.
   *
   * @param {string?} [remote] - VCS-friendly repository url or reference
   */
  constructor(remote?: string) {
    if (!remote) {
      const { repository } = new PackageManifest().pkg
      remote = get(repository, 'url', <string>repository)
    }

    /**
     * Regular expression used to extract repository name, owner, and provider.
     *
     * @see https://regex101.com/r/pjJWeg
     *
     * @const {RegExp} regex
     */
    const regex: RegExp =
      /(?:(?:https?:\/{2})?(?<provider>bitbucket|git(?:hub|lab))(?:(?:\.(?:com|org)\/)|:))?(?<owner>[\w-]+)\/(?<repo>[\w-]+)/

    // extract repository data
    const {
      owner,
      provider,
      repo
    } = defaults(get(regex.exec(remote = trim(remote)), 'groups', {}), {
      owner: '',
      provider: ifelse(regex.test(remote), RepositoryProvider.GITHUB, ''),
      repo: ''
    })

    // set repository owner, provider, name
    this.owner = owner
    this.provider = <RepositoryProvider>provider
    this.repo = repo
  }

  /**
   * Get `this` repository provider hostname.
   *
   * @see {@linkcode RepositoryHost}
   *
   * @public
   *
   * @return {RepositoryHost} Repository hostname
   */
  public get host(): RepositoryHost {
    return <RepositoryHost>template('{provider}.{tld}', {
      provider: this.provider,
      tld: ifelse(
        this.provider === RepositoryProvider.BITBUCKET,
        'org',
        'com'
      )
    })
  }

  /**
   * Get keywords used in reference URLs.
   *
   * @see {@linkcode RepositoryKeywords}
   *
   * @public
   *
   * @return {RepositoryKeywords} Repository keywords object
   */
  public get keywords(): RepositoryKeywords {
    switch (this.provider) {
      case RepositoryProvider.BITBUCKET:
        return { commit: 'commit', issue: 'issues', pr: 'pull-requests' }
      case RepositoryProvider.GITLAB:
        return { commit: 'commit', issue: 'issues', pr: 'merge_requests' }
      default:
        return { commit: 'commit', issue: 'issues', pr: 'pull' }
    }
  }

  /**
   * Get `this` repository url.
   *
   * @public
   *
   * @return {string} Repository URL
   */
  public toString(): string {
    return template('https://{host}/{owner}/{repo}', cast(this))
  }
}

export default Repository
