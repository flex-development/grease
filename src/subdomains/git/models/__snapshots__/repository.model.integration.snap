// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:git/models/Repository > validation > should fail validation if schema is invalid 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "minLength": "owner must be longer than or equal to 1 characters",
    },
    "property": "owner",
    "value": "",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isEnum": "provider must be one of the following values: bitbucket, github, gitlab",
    },
    "property": "provider",
    "value": "",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "minLength": "repo must be longer than or equal to 1 characters",
    },
    "property": "repo",
    "value": "",
  },
]
`;
